### PR TITLE
New qualification status page + random stuff

### PR DESCRIPTION
**Pull request by krzys-h**
(Sorry for putting so much stuff into one PR but they are all somewhat dependent on each other and I'm way too tired of rebasing changes all the time)

The main focus of this pull request is a new qualification status page which shows:
* Your current qualification status (accepted/rejected/cancelled as participant/lecturer)
* For participants: if you don't have a status yet, info what you are missing:
  * "You didn't select any workshops"
  * "You didn't fill out the cover letter"
  * "You did everything right, now go ahead and solve the tasks and wait for the results" (may change when we introduce solution uploads)
* For participants: the table with qualification results (moved from the program page - that was quite an awkward place to have it)
* For lecturers: the table with your workshops (moved from the 'your workshops' page - the big DataTable was useless for normal lecturers who only have one or two lectures in a year) with new, clear information on what data is missing:
  * "Fill out your public page"
  * "Upload the qualification tasks"
  * "Set the qualification threshold"
* Historical data

I additionally sneaked in some general improvements:
* The profile editor is properly separated into multiple pages (#130)
* All of the tabs of the editor now have short descriptions - things like "what is the profile page for" may not have been entirely clear before
* If you are a lecturer, and didn't fill out your profile page, the workshop editor will show a hint that it may be a good idea to do so
* The workshop list is now separated into multiple pages per year (#86) and loads faster because I fixed some prefetch_related issues
* The "Warsztaty" and "Profil" menu items were merged into "Zgłoś się!" - the drop-down on "Warsztaty" with only one entry (for normal people) looked a bit awkward
* Added caching for `Camp.objects.latest()`
* Fixed the workshop qualification buttons appearing even if the workshop is no longer editable (is from previous edition)

screencapture-localhost-8000-me-status-2021-01-01-20_54_20

---
This change is [Reviewable](https://reviewable.io/reviews/warsztatywww/aplikacjawww/269)

### CODE REVIEW

**Review by cytadela8 — CHANGES_REQUESTED** [review]
.

**Review by cytadela8 — COMMENTED** [review]
Reviewable doesn't provide any way to reset comments after first or change back to reviewing each commit separately after first comment is made so I will review this PR probably fully on Github....

> *[Reviewable](https://reviewable.io/reviews/warsztatywww/aplikacjawww/269#-:-MQ7EUwG9ZWtp52OWiFE:b-yeupus)* status: 0 of 30 files reviewed, 3 unresolved discussions (waiting on @krzys-h)

File: templates/_workshopStatusTable.html [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      <th scope="row" style="width: 50%">
+        <a href="{% url 'workshop_edit' workshop.year.pk workshop.name %}">
+          {{ workshop.title }}
+        </a>
+      </th>
+      <td style="width: 50%">
+        {% if workshop.status == 'Z' %}
+          <span class="qualified"> {{ workshop.get_status_display }} </span>
+          {% if workshop.page_content_is_public %}
+            <br><span class="waiting-qualified">Uzupełnij stronę warsztatów!</span>
```
Comment (cytadela8) [inline templates/_workshopStatusTable.html:22]: This doesn't really make sense on old workshops when you can't feel in forms any more.

File: wwwapp/models.py [hunk 2]
```diff
@@ -42,14 +43,23 @@ def are_proposals_open(self) -> bool:
...
 
     def is_qualification_editable(self) -> bool:
         if self.start_date:
             return self.are_workshops_editable() and datetime.datetime.now().date() < self.start_date
         else:
             return self.are_workshops_editable()
 
+    def save(self, *args, **kwargs):
+        super().save(*args, **kwargs)
+        Camp.current.cache_clear()
```
Comment (cytadela8) [inline wwwapp/models.py:56]: This will not work on production! We have multiple workers (processes).

**Review by cytadela8 — CHANGES_REQUESTED** [review]

File: templates/_workshopStatusTable.html [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    <tr>
+      <th scope="row" style="width: 50%">
+        <a href="{% url 'workshop_edit' workshop.year.pk workshop.name %}">
+          {{ workshop.title }}
+        </a>
+      </th>
+      <td style="width: 50%">
+        {% if workshop.status == 'Z' %}
+          <span class="qualified"> {{ workshop.get_status_display }} </span>
+          {% if workshop.page_content_is_public %}
```
Comment (cytadela8) [inline templates/_workshopStatusTable.html:21]:
```suggestion
          {% if not workshop.page_content_is_public %}
```
Logic is reversed

File: templates/mydata_status.html [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+          <div style="font-size: 300%;"><span class="maybe-qualified">Odwołano przyjazd</span></div>
+          <div>
+            {% if gender == 'F' %}
+              Zostałaś zakwalifikowana na warsztaty, ale odwołałaś swój przyjazd.
+            {% else %}
+              Zostałeś zakwalifikowany na warsztaty, ale odwołałeś swój przyjazd.
+            {% endif %}
+          </div>
+        {% elif not current_status.status %}
+          {% if not has_cover_letter %}
```
Comment (cytadela8) [inline templates/mydata_status.html:62]: We need indication of profile info missing, but I would leave this for forms rework.

File: wwwapp/views.py [hunk 5]
```diff
@@ -281,21 +341,31 @@ def workshop_edit_view(request, year=None, name=None):
     workshop_url = workshop_url.split('SOMENAME')
     workshop_url[0:1] = workshop_url[0].split('9999')
 
+    profile_warnings = []
+    if not workshop or workshop.lecturer.filter(user=request.user).exists():  # The user is one of the lecturers for this workshop
+        if len(request.user.userprofile.profile_page) <= 50:  # The user does not have their profile page filled in
+            profile_warnings.append(Template("""
+                    <strong>Nie uzupełnił{% if user.userprofile.gender == 'F' %}aś{% else %}eś{% endif %} swojej
+                    <a href="{% url 'mydata_profile_page' %}">strony profilowej</a>.</strong>
```
Comment (cytadela8) [inline wwwapp/views.py:349]:
```suggestion
                    <a target="_blank" href="{% url 'mydata_profile_page' %}">strony profilowej</a>.</strong>
```

Also maybe add similar thing to status page

**Review by cytadela8 — CHANGES_REQUESTED** [review]
More comments.

The colours and styling are not ideal. I think we should leave it as is due to size of this PR and I will try to improve the styling in a separate PR. I also want to move around the links in menu. Currently "Zapisz się" kind of contains 3 identical links... + 1 workshop registration link.

File: templates/mydata_status.html (hunk 4 above)
Comment (krzys-h) [reply templates/mydata_status.html:62]: I know, that's why I didn't touch it for now

**Review by cytadela8 — CHANGES_REQUESTED** [review]
> Reviewed 30 of 30 files at r1, 2 of 7 files at r3, 4 of 8 files at r4, 8 of 11 files at r5, 5 of 6 files at r8, 4 of 4 files at r9, 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/warsztatywww/aplikacjawww/269)* status: all files reviewed, 2 unresolved discussions (waiting on @krzys-h)

___
*[templates/mydata_status.html, line 15 at r10](https://reviewable.io/reviews/warsztatywww/aplikacjawww/269#-MQ85bP-DA49Uaeajggu:-MQ85bP-DA49Uaeajggv:b-bvdh2u) ([raw file](https://github.com/warsztatywww/aplikacjawww/blob/be466dac661f369662c7e4a4392496113d861642/templates/mydata_status.html#L15)):*
> ```HTML
>       </div>
>     {% else %}
>       {% if current_status.type == 'lecturer' %}
> ```

Minor, Can we get a notification to fill in profile page here?
___
*[wwwapp/models.py, line 56 at r10](https://reviewable.io/reviews/warsztatywww/aplikacjawww/269#-MQ84rYV4IVZ5UpSnAP7:-MQ84rYV4IVZ5UpSnAP8:bsup8lu) ([raw file](https://github.com/warsztatywww/aplikacjawww/blob/be466dac661f369662c7e4a4392496113d861642/wwwapp/models.py#L56)):*
> ```Python
>     def save(self, *args, **kwargs):
>         super().save(*args, **kwargs)
>         Camp.current.cache_clear()
> ```

:warning: This won't work on production as we have multiple processes there and this cache is per process!

**Review by krzys-h — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/warsztatywww/aplikacjawww/269)* status: 20 of 31 files reviewed, 2 unresolved discussions (waiting on @cytadela8)

___
*[templates/mydata_status.html, line 15 at r10](https://reviewable.io/reviews/warsztatywww/aplikacjawww/269#-MQ85bP-DA49Uaeajggu:-MQ8E3xcAFmqZ_8OtmMx:b-ek72f9) ([raw file](https://github.com/warsztatywww/aplikacjawww/blob/be466dac661f369662c7e4a4392496113d861642/templates/mydata_status.html#L15)):*
<details><summary><i>Previously, cytadela8 (Artur Puzio) wrote…</i></summary><blockquote>

Minor, Can we get a notification to fill in profile page here?
</blockquote></details>

The profile page is not technically a hard requirement and everything else on this page is a hard requirement. They already get a notification on the workshop registration page before they apply, I feel like that's enough.
___
*[wwwapp/models.py, line 56 at r10](https://reviewable.io/reviews/warsztatywww/aplikacjawww/269#-MQ84rYV4IVZ5UpSnAP7:-MQ8EX1nDseRd2W4R3Am:b-886gsx) ([raw file](https://github.com/warsztatywww/aplikacjawww/blob/be466dac661f369662c7e4a4392496113d861642/wwwapp/models.py#L56)):*
<details><summary><i>Previously, cytadela8 (Artur Puzio) wrote…</i></summary><blockquote>

:warning: This won't work on production as we have multiple processes there and this cache is per process!
</blockquote></details>

Fixed

**Review by cytadela8 — APPROVED** [review]
:duck: 

> Reviewed 10 of 10 files at r11, 5 of 5 files at r12, 4 of 4 files at r13, 6 of 6 files at r14, 5 of 5 files at r15, 1 of 1 files at r16, 1 of 1 files at r17.
> *[Reviewable](https://reviewable.io/reviews/warsztatywww/aplikacjawww/269#-:-MQ8FDT3D7FtMnMh-FFD:bcmjph0)* status: :shipit: complete! all files reviewed, all discussions resolved

___
*[templates/mydata_status.html, line 15 at r10](https://reviewable.io/reviews/warsztatywww/aplikacjawww/269#-MQ85bP-DA49Uaeajggu:-MQ8EzoZAp89LvvoEVHD:b32wtd) ([raw file](https://github.com/warsztatywww/aplikacjawww/blob/be466dac661f369662c7e4a4392496113d861642/templates/mydata_status.html#L15)):*
<details><summary><i>Previously, krzys-h (krzys_h) wrote…</i></summary><blockquote>

The profile page is not technically a hard requirement and everything else on this page is a hard requirement. They already get a notification on the workshop registration page before they apply, I feel like that's enough.
</blockquote></details>

ok

**Pull request merged**